pubsub/driver/driver.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package driver defines a set of interfaces that the pubsub package uses to
+// interact with the underlying pubsub services.
+package driver
+
+import "context"
+
```

Comment:
OK to do in a separate PR if you prefer, but you should add `As` functions for `Topic`, `Subscription`, and `Message`. See `blob.Driver` for examples.

`Message` is used for both reading and writing, which makes it interesting for `As`. An `As` function ```As(i interface{}) bool``` makes sense for reading, but for writing you want to give access to the underlying type of whatever message is going to be sent to the server. In `blob`, we do this via `WriterOptions.BeforeWrite`. So you could:

1. Add a `BeforeWrite`-like function to `Message`, that's only used when writing
2. Split `Message` into separate types for reading (with `As`) and writing (with `BeforeWrite`)
3. Add an `Options` struct with `BeforeWrite` to `SendBatch`.

Reply:
Added issue #657 .